lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchEvaluatorShimManager.java

```diff
@@ -0,0 +1,339 @@
+
+    this.jobJarMaker = jobJarMaker;
+    this.launchCommandBuilder = launchCommandBuilder;
+
+    this.azureBatchAccountKey = azureBatchAccountKey;
+    this.azureBatchAccountName = azureBatchAccountName;
+    this.azureBatchAccountUri = azureBatchAccountUri;
+
+    this.outstandingResourceRequests = new ConcurrentHashMap<>();
+    this.activeResources = new ConcurrentHashMap<>();
+
+    this.evaluatorShimCommandChannel = remoteManager
+        .registerHandler(EvaluatorShimProtocol.EvaluatorShimStatusProto.class, this);
+  }
+
+  public void onResourceRequested(final String containerId, final ResourceRequestEvent resourceRequestEvent) {
+    try {
+      createAzureBatchTask(containerId);
+      this.outstandingResourceRequests.put(containerId, resourceRequestEvent);
+      this.updateRuntimeStatus();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void onNext(final RemoteMessage<EvaluatorShimProtocol.EvaluatorShimStatusProto> statusMessage) {
+
+    EvaluatorShimProtocol.EvaluatorShimStatusProto message = statusMessage.getMessage();
+    String containerId = message.getContainerId();
+    String remoteId = message.getRemoteIdentifier();
+
+    LOG.log(Level.INFO, "Got a status message from evaluator shim = {0} with containerId = {1}.",
+        new String[] {remoteId, containerId});
+
+    synchronized (this.outstandingResourceRequests) {
+      ResourceRequestEvent resourceRequestEvent = AzureBatchEvaluatorShimManager.this
+          .outstandingResourceRequests
+          .get(containerId);
+
+      if (resourceRequestEvent == null) {
+        LOG.log(Level.WARNING, "Received an evaluator shim status message from an unknown "
+            + "evaluator shim. Container id = {0}, remote id = {1}.", new String[] {containerId, remoteId});
+      } else {
+        LOG.log(Level.FINEST, "Notifying REEF of a new node: {0}", remoteId);
+        this.reefEventHandlers.onNodeDescriptor(NodeDescriptorEventImpl.newBuilder()
+            .setIdentifier(RemoteIdentifierParser.parseNodeId(remoteId))
+            .setHostName(RemoteIdentifierParser.parseIp(remoteId))
+            .setPort(RemoteIdentifierParser.parsePort(remoteId))
+            .setMemorySize(resourceRequestEvent.getMemorySize().get())
+            .build());
+
+        LOG.log(Level.FINEST, "Firing a new ResourceAllocationEvent for remoteId = {0}.", remoteId);
+        this.reefEventHandlers.onResourceAllocation(
+            ResourceEventImpl.newAllocationBuilder()
+                .setIdentifier(containerId)
+                .setNodeId(RemoteIdentifierParser.parseNodeId(remoteId))
+                .setResourceMemory(resourceRequestEvent.getMemorySize().get())
+                .setVirtualCores(resourceRequestEvent.getVirtualCores().get())
+                .setRuntimeName(RuntimeIdentifier.RUNTIME_NAME)
+                .build());
+      }
+
+      this.outstandingResourceRequests.remove(containerId);
```

Comment:
If I understand correctly, this is actually an ack message from evaluator to driver for node allocation. Records will be removed from outstandingResourceRequests, if it is acked.

When resourceRequestEvent is null (Line 155), I don't think we should add the container in activeResources(Line 171)

Reply:
Yes, you're absolutely right. I moved it into the "else" block.